### ISSUE TITLE
Support UTF8 passwords

### ISSUE THREAD

**Issue by jesperpedersen** [agroal/pgagroal#51]
Requires updates to SASL prep

**Comment by fluca1978** [comment]
As far as I can tell, `strlen` seems to support multibytes strings (I thought `wcslen` and friends).
Apparently the following code snippet works finr on my machine, but I'm sure I'm missing something:


```
#include <stdio.h>
#include <string.h>
#include <stdlib.h>


static int
sasl_prep(char* password, char** password_prep)
{
    int length = strlen( password );
    *password_prep = (char*) malloc( length );
    strncpy( *password_prep, password, length);

    return 0;
}

#define SIZE 64
#define PWD1 "I ❤ pgagroal, the 🐘 PostgreSQL connection pooler! 🎉"
#define PWD2 "LÙCA FERRÀRÌ"

void
main()
{
    char *src_password, *dst_password;

    src_password = (char*) malloc( SIZE );
    dst_password = (char*) malloc( SIZE );
    memset( src_password, 0, SIZE );
    memset( dst_password, 0, SIZE );

    memcpy( src_password, PWD1 , sizeof( PWD1 ) );
    printf( "\nOriginal : [%s]\n", src_password );
    sasl_prep( src_password, &dst_password );
    printf( "\nSASL : [%s]\n", dst_password );

    memcpy( src_password, PWD2 , sizeof( PWD2 ) );
    printf( "\nOriginal : [%s]\n", src_password );
    sasl_prep( src_password, &dst_password );
    printf( "\nSASL : [%s]\n", dst_password );

}
```

If that is right, the only need is to change `strdup` to `strncpy`. Probably there shold be something else to check for longer strings (overflows?).
This is surely not my area of experise, so forgive me in case the above is totally wrong.

**Comment by jesperpedersen** [comment]
Unfortunately it isn't as simple.

See section 2.2 in https://datatracker.ietf.org/doc/html/rfc5802 for the overall requirements (`Normalize(str)`). It requires an implementation of 

* https://datatracker.ietf.org/doc/html/rfc4013
* https://datatracker.ietf.org/doc/html/rfc3454
* https://datatracker.ietf.org/doc/html/rfc3629

These checks needs to be implemented in `admin.c` as well.

**Comment by fluca1978** [comment]
Definitely something out of my expertise.
Any chance we can reuse PostgreSQL code here? <https://github.com/postgres/postgres/blob/master/src/common/saslprep.c#L1044>

**Comment by jesperpedersen** [comment]
It ihas to be a clean-room implementation - although you can look at it if you give credit.

We can't have a dependency on `postgresql-devel`.

**Comment by jesperpedersen** [comment]
@Userfrom1995 Can you take this ?

**Comment by Userfrom1995** [comment]
> [@Userfrom1995](https://github.com/Userfrom1995) Can you take this ?

Sure, I can take this one.

**Comment by jesperpedersen** [comment]
@Userfrom1995 Can we close this ?

**Comment by Userfrom1995** [comment]
> [@Userfrom1995](https://github.com/Userfrom1995) Can we close this ?

Yes, I think we can close it if you are satisfied with the implementation.
Regarding the normalization and the SASL prep discussion we had, I think if we ever need it, I’ll handle it in a separate PR with a dedicated issue. 
I believe most users won’t encounter that problem anyway, as most users don’t use such complicated passwords.

**Comment by jesperpedersen** [comment]
@Userfrom1995 Its part of the #52 issue, but we can handle it in that issue. Thanks !
